### PR TITLE
Chapter 2 solutions

### PR DESCRIPTION
**Pull request by NanthR**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -620,7 +624,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven [x] = [x]
+takeEven (x:_:xs) = x : takeEven xs
```
Comment (vrom911) [inline src/Chapter2.hs:630]: Great pattern matching skills 💯

File: src/Chapter2.hs [hunk 2]
```diff
@@ -727,7 +734,8 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate [] = []
+smartReplicate (x:xs) = replicate x x ++ smartReplicate xs
```
Comment (vrom911) [inline src/Chapter2.hs]: This is right! 💯 

You can notice that what is happening is you walk through each element of the list, change it to something else (by using `replicate` function) and then concatenate all of the results into one list.

In Haskell, there is the `map` function that does exactly the first part of it (walk through each element of the list, change it to something else ).

After using `map` you will get a list of lists. To flatten such a list, you can use the `concat` function:

```haskell
ghci> concat [[3, 1, 2], [4, 0, 1]]
[3,1,2,4,0,1]
```

This pattern of using `map` and `concat` is very common, so the standard library has even a standard function `concatMap` which is a combination of `map` and `concat` (as the name suggests).

File: src/Chapter2.hs [hunk 3]
```diff
@@ -841,7 +851,13 @@ list.
...
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [Int] -> [Int]
+rotate n l
+  | n < 0 = []
+  | null l = []
+  | otherwise = 
+      let len = length l
+      in take len (drop (mod n len) (cycle l)) -- mod because a rotate of len just returns the same thing
```
Comment (vrom911) [inline src/Chapter2.hs]: Really nice catch, it makes the implementation way more efficient! 👍🏼 

You can also add another trick to speed up – return the list itself if `mod` is 0 😉

File: src/Chapter2.hs [hunk 4]
```diff
@@ -857,7 +873,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [Int] -> [Int]
+rewind [] = []
+rewind (x:xs) = rewind xs ++ [x]
```
Comment (vrom911) [inline src/Chapter2.hs]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by vrom911 — APPROVED** [review]
Fantastic job! 👏🏼 
I really enjoyed your neat and thoughtful solutions :+1:

File: src/Chapter2.hs (hunk 1 above)
Comment (NanthR) [reply src/Chapter2.hs:630]: Yes, thank you

File: src/Chapter2.hs (hunk 2 above)
Comment (NanthR) [reply src/Chapter2.hs]: Oh, ok. I did not know that. Will look into that

File: src/Chapter2.hs (hunk 4 above)
Comment (NanthR) [reply src/Chapter2.hs]: Yeah, I read about them but didn't properly know how to use them. I have looked around more and have become a bit more confident in using them. And, thanks for the suggestion

**Pull request merged**